### PR TITLE
fix unit test TestValidatePersistentVolumeClaimUpdate

### PR DESCRIPTION
**Pull request by CaoShuFeng**
Fixes: #56393 

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: pkg/apis/core/validation/validation_test.go [hunk 1]
```diff
@@ -1420,7 +1420,7 @@ func TestValidatePersistentVolumeClaimUpdate(t *testing.T) {
 		toggleBlockVolumeFeature(scenario.enableBlock, t)
 		scenario.oldClaim.ResourceVersion = "1"
 		scenario.newClaim.ResourceVersion = "1"
-		errs := ValidatePersistentVolumeClaimUpdate(scenario.newClaim, scenario.oldClaim)
```
Comment (CaoShuFeng) [inline pkg/apis/core/validation/validation_test.go]: Ref:
https://github.com/kubernetes/kubernetes/blob/e7ad6e60081887d67b987cb0429b52c5a4ff878c/pkg/apis/core/validation/validation.go#L1819

newClaim is updated by function PersistentVolumeClaimUpdate(). So I passed the copied one into it. So test cases will not effect each other.

**Comment by CaoShuFeng** [comment]
/assign @liggitt 
for approve

File: pkg/apis/core/validation/validation_test.go (hunk 1 above)
Comment (deads2k) [inline pkg/apis/core/validation/validation_test.go]: Validation should not be mutating an object.

Comment (liggitt) [reply pkg/apis/core/validation/validation_test.go]: ValidateObjectMetaUpdate, I think

Comment (gnufied) [reply pkg/apis/core/validation/validation_test.go]: The status copy over to new pvc has been there forever. I dunno if we can safely remove it. cc @saad-ali

Comment (gnufied) [reply pkg/apis/core/validation/validation_test.go]: > ValidateObjectMetaUpdate, I think
 
@liggitt I am not sure I understand.

**Comment by deads2k** [comment]
I'm not seeing how this fixing the problem.  The referenced code is a different bug and this change is protecting an unrelated field which near as I can tell won't race as compared to https://github.com/kubernetes/kubernetes/pull/57199 which fixes a demonstrable race against a shared global.

**Comment by deads2k** [comment]
> I'm not seeing how this fixing the problem. The referenced code is a different bug and this change is protecting an unrelated field which near as I can tell won't race as compared to #57199 which fixes a demonstrable race against a shared global.

It could be re-use of an existing API object.  I guess we'll see.

**Comment by gnufied** [comment]
@deads2k it is fixing https://github.com/openshift/origin/issues/17769 because what is happening is - `ValidatePersistentVolumeClaimUpdate` is mutating `newPvc.Status` and replacing it with `oldPvc.Status`.  What that is doing is - `validClaim` object that is supposed to have phase as "Bound"  - that phase gets removed by something like - https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/validation/validation_test.go#L1355 

So, when phase of `validClaim` is not bound, we can't edit the size anymore and that causes the test to fail. The failure depends on order in which each test is run.

**Comment by liggitt** [comment]
> ValidatePersistentVolumeClaimUpdate is mutating newPvc.Status and replacing it with oldPvc.Status

So it is. That's a bug. That should happen in the strategy, not in validation.

**Comment by deads2k** [comment]
> @deads2k it is fixing openshift/origin#17769 because what is happening is - ValidatePersistentVolumeClaimUpdate is mutating newPvc.Status and replacing it with oldPvc.Status. What that is doing is - validClaim object that is supposed to have phase as "Bound" - that phase gets removed by something like - https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/validation/validation_test.go#L1355
> 
> So, when phase of validClaim is not bound, we can't edit the size anymore and that causes the test to fail. The failure depends on order in which each test is run.

Ok.  And we think this is the correct fix instead of doing the work in `PrepareForUpdate`?  Seems unlikely.

**Comment by deads2k** [comment]
You should probably stop mutating the oldPVC as part of your validation method too.  Removing https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/validation/validation.go#L1778-L1783

**Comment by gnufied** [comment]
@deads2k @liggitt not that I disagree but the commit that introduced mutation is - https://github.com/kubernetes/kubernetes/commit/95bd170ca26c82f3df51ccf8ac08a827eff77ba6#diff-f19a7777a36882029db668a427f950feR432

I am in favour of dropping the mutation but I am not sure if something depends on it.

**Comment by deads2k** [comment]
> @deads2k @liggitt not that I disagree but the commit that introduced mutation is - 95bd170#diff-f19a7777a36882029db668a427f950feR432
> 
> I am in favour of dropping the mutation but I am not sure if something depends on it.

There's only one caller, in the strategy.

File: pkg/apis/core/validation/validation.go [hunk 2]
```diff
@@ -1816,7 +1816,6 @@ func ValidatePersistentVolumeClaimUpdate(newPvc, oldPvc *core.PersistentVolumeCl
 		allErrs = append(allErrs, ValidateImmutableField(newPvc.Spec.VolumeMode, oldPvc.Spec.VolumeMode, field.NewPath("volumeMode"))...)
 	}
 
-	newPvc.Status = oldPvc.Status
```
Comment (CaoShuFeng) [inline pkg/apis/core/validation/validation.go]: > So it is. That's a bug. That should happen in the strategy PrepareForCreate/PrepareForUpdate methods, not in validation.

This is already done here:
https://github.com/kubernetes/kubernetes/blob/6c5f0799bdd65ca99394017cb6db54bcc1464961/pkg/registry/core/persistentvolumeclaim/strategy.go#L75

So we can just remove it here.

**Comment by CaoShuFeng** [comment]
> You should probably stop mutating the oldPVC as part of your validation method too. Removing https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/validation/validation.go#L1778-L1783

This will tighten the update API, and make newPvc.Spec.VolumeName become immutable

**Pull request closed without merging**